src/script/ismine.h

```diff
@@ -9,22 +9,44 @@
 #include <script/standard.h>
 
 #include <stdint.h>
+#include <bitset>
 
 class CKeyStore;
 class CScript;
 
 /** IsMine() return codes */
 enum isminetype
 {
-    ISMINE_NO = 0,
-    ISMINE_WATCH_ONLY = 1,
-    ISMINE_SPENDABLE = 2,
-    ISMINE_ALL = ISMINE_WATCH_ONLY | ISMINE_SPENDABLE
+    ISMINE_NO         = 0,
+    ISMINE_WATCH_ONLY = 1 << 0,
+    ISMINE_SPENDABLE  = 1 << 1,
+    ISMINE_ALL        = ISMINE_WATCH_ONLY | ISMINE_SPENDABLE,
+    ISMINE_ENUM_ELEMENTS,
 };
 /** used for bitflags of isminetype */
 typedef uint8_t isminefilter;
 
 isminetype IsMine(const CKeyStore& keystore, const CScript& scriptPubKey);
 isminetype IsMine(const CKeyStore& keystore, const CTxDestination& dest);
 
+struct CachableAccount
+{
+    enum class Type : uint8_t {
```

Comment:
The code is this way because I specifically objected to passing the CacheType to GetCacheable, which is how it was written before, because the CacheType _is_ in fact a property of the CachableAccount object: any given such object is always either a credit account or a debit account, and it is _always_ an error to call GetCachable with any CacheType other than the correct one for the particular CachableAccount object. Associating the type with the object prevents this error.

What I thought you were getting at in the first paragraph was that the type of a CachableAccount is a static property that is known at compile time. This I agree with, and you could certainly make it a type parameter, although I think that's unusually esoteric for C++ code, and making it an ordinary member is perfectly defensible. But I think not storing it at all, and leaving it implicit, is not defensible.

Reply:
@gwillen it sounds like you are worried about an error that could have occurred in a previous or hypothetical version of the code where GetCacheable function took both CachableAmount and Type arguments. I agree that this would be more error prone than the current code, but it's not what I'm suggesting. 

The current code avoids one problem (type being inconsistent with CachableAmount variable) but stores extra runtime data for no reason and adds a weird action-at-distance thing where you have to look 3 different places to see what GetCacheable will do (call site, function definition, cache variable construction). I don't there is anything wrong with actual interface I suggested:

```c++
enum CacheType { DEBIT, CREDIT, IMMATURE_CREDIT, AVAILABLE_CREDIT };
CAmount GetCachable(CacheType type, const isminefilter& filter, bool recalculate) const;
```

If you don't like it for some reason, that's fine with. I already acked this PR in its current form. But if you want to call my suggestion "not defensible" you should be able to point out some actual error or mistake or problem that could it could cause.

Reply:
It was not hypothetical, it was indeed the previous version of the code, which I think you can see in the history of the PR (but maybe a rebase wiped it out.) I responded too quickly because I was frustrated that you appeared to be suggesting to go back to a previous version, without addressing my comments discussing why I was opposed to the previous version. In fact I did misunderstand your suggestion, and the thing you actually suggested makes sense to me. Sorry for responding the way I did.

Reply:
@ryanofsky That sounds reasonable, but how were you envisioning accessing the corresponding cache type from inside `GetCachable`? I started rewriting the code and ended up with
```diff
-    mutable CachableAccount m_debit{CachableAccount::Type::DEBIT};
-    mutable CachableAccount m_credit{CachableAccount::Type::CREDIT};
-    mutable CachableAccount m_credit_immature{CachableAccount::Type::CREDIT};
-    mutable CachableAccount m_credit_available{CachableAccount::Type::CREDIT};
+    mutable CachableAccount accounts[CACHETYPE_ENUM_ELEMENTS];
```
which seems like it will degrade readability.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/15780/files#r275588760 from kallewoof

> I started rewriting the code and ended up with

Yes, I would probably implement this with an array. But a smaller change would be to use a conditional:

```c++
auto& cache = type == CREDIT ? m_credit : type == DEBIT ? m_debit : ...;
```

> which seems like it will degrade readability.

Where specifically do you think this decreases readability?

- Readability outside the GetCached function is improved by this suggestion because it replaces a complicated in/out parameter with a simple enum mode argument.
- Readability inside the GetCached function should be improved even if the code is longer by a line or two because a layer of indirection is gone and you can see the actual objects that are being read and written to instead of having to trace through aliases.
- Overall readability should be improved because instead of cache struct instances being accessed from many parts of wallet code they are accessed only in one place.